core/domain/query_jobs_one_off.py

```diff
@@ -0,0 +1,143 @@
+# coding: utf-8
+#
+# Copyright 2016 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Jobs to execute and get result of a query."""
+import ast
+import datetime
+
+from core import jobs
+from core.domain import exp_services
+from core.domain import query_domain
+from core.platform import models
+
+(user_models, job_models, exp_models) = (
+    models.Registry.import_models(
+        [models.NAMES.user, models.NAMES.job, models.NAMES.exploration]))
+
+
+class QueryOneOffJob(jobs.BaseMapReduceJobManager):
+    """One-off job for excuting query with given query parameters.
+    For each user we check if he/she satisfies query criteria. If user satisfies
+    query criteria then yeilds a tuple of (query_id, user_id). Reducer function
+    stores all such satisfying user_ids in query result model."""
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [user_models.UserSettingsModel]
+
+    @staticmethod
+    def map(user_settings_model):
+        query_id = jobs.BaseMapReduceJobManager.get_mapper_param('query_id')
+        query = job_models.QueryModel.get(query_id)
+        query_parameters = query_domain.get_parameters_from_query(query)
+        user_id = user_settings_model.id
+        user_qualified = True
+
+        if user_id == query.submitter_id:
+            user_qualified = False
+
+        if user_settings_model.username == 'tmpsuperadm1n':
+            user_qualified = False
+
+        if (query_parameters.created_or_edited_exploration_in_n_days is not None
+                and user_qualified):
+            days = query_parameters.created_or_edited_exploration_in_n_days
+            created_exploration_in_n_days = False
+            edited_exploration_in_n_days = False
+
+            user_contributions = user_models.UserContributionsModel.get(user_id)
+            created_explorations = exp_services.get_multiple_explorations_by_id(
+                user_contributions.created_exploration_ids)
+
+            edited_explorations = (
```

Comment:
Yeah. Your suggestion is totally fine. The issue with it is that such data will take time to build up. Initially it will be set to None for all users. Over the time people will come and edit or create explorations and those attributes will be changed. 

The main reason we need query is to inform old users about latest changes in Oppia. That would not happen since old users may not visit the site at all and if they do visit then we do not need to send email to them. So even if we add new attributes I don't get how we can use them here ?